### DISCUSSION TITLE
Should I put AppContext to layout file ?

### DISCUSSION THREAD

**Discussion by hungcrush** [vercel/next.js#49352] [Help]
### Summary

In my app, I need to store global data such as settings and user information. To load this data from the server-side, I created an AppContext and added it to the layout.tsx file. Is this a good approach, or do you have any better suggestions?

### Additional information
I have read this:

> You should remove "use client" from child.tsx.

> A "use client" directive means that component is a boundary between server and client components. Since parent.tsx is already inside the client boundary, everything imported will be client components already. And that includes child.tsx.

> Why there's this warning? Since it's a boundary, you can potentially use it in a server component. However it's not allowed to have props that can't be serialized in that case. If you import child.tsx and render it inside a server component, there's no way to pass a function as the prop.

> So this warning also suggests that you should not add "use client" to every client component, but only those boundaries that are directly used in server components.

**Comment by SamKomesarook** [comment]
In our case, we also require a global context that mostly all pages require, and we have wrapped the top-level-layout children in the context provider.

In the future, we will, where possible, move the providers further down the component tree so that only the components that need them are wrapped with them.

One alternative is in the [ecommerce demo](https://github.com/vercel/commerce), they have an interesting way of saving the cart state, which is stored in the backend (in this case, a shopify headless storefront). That alleviates the need for a react context for that variable.

However, I am not sure what the performance tradeoff is for having to make the remote state request every so often instead of a react context. Plus, if you don't have a backend or the one you have doesn't offer that kind of state, then that's another holdup...

**Reply by hungcrush** [reply]
Thank you so much
